[ch13440] Add ACN Validation

gungdesurya1 opened the pull request:
Add ACN Validation with format nnn nnn nnn as a number

The form will shows no error and the submit button will enable like the image below
![Screen Shot 2021-03-16 at 08 41 21](https://user-images.githubusercontent.com/79816724/111239459-30cb8e00-8634-11eb-9d4b-53cbeae55926.png)

Error message will appear if users put the wrong number like the image below
![Screen Shot 2021-03-16 at 08 41 41](https://user-images.githubusercontent.com/79816724/111239530-5193e380-8634-11eb-985c-c73712d079b5.png)

And if users put non numeric or the digits is less than 9 or the format is not `nnn nnn nnn` or `nnnnnnnnn`, error message will be appear
![Screen Shot 2021-03-16 at 08 42 11](https://user-images.githubusercontent.com/79816724/111239683-a6375e80-8634-11eb-8266-c89764416c5a.png)

Commit:
setup dev environment [ch13440]

Commit:
Add ACN validation algorithm to the input text [ch13440]

- add basic html form and error message

Commit:
Add styling to form [13440]

Commit:
Add unit testing (karma-jasmin) [ch13440]

ellyxc commented on index.js:
```diff
@@ -0,0 +1,33 @@
+    }
+  };
+
+  $scope.onChangeAcn = function (acn) {
+    const pattern = /^(\d{3} \d{3} \d{3}|\d{9})$/i;
+    const isFormatValid = acn.match(pattern) ? true : false;
+    $scope.form.acn.$setValidity("format", isFormatValid);
+    if (!isFormatValid) return false;
+
+    acn = acn.replaceAll(" ", "");
+    let totalWeight = 0;
+    let weight = 8;
+    for (let i = 0; i < acn.length - 1; i++) {
+      const digit = acn[i];
+      totalWeight += digit * weight;
+      weight--;
+    }
+
+    const remainder = totalWeight % 10;
+    let complement = 10 - remainder;
+    if (complement == 10) complement = 0;
+
+    const isACNValid = acn[8] == complement;
+    $scope.form.acn.$setValidity("acnValid", isACNValid);
+    return isACNValid;
+  };
```
Hi Gung De, this is looking really good.

a feedback on this function, it is not about the logic or output but this is about separation of concern, in MVC, controller accepts input and converts it to commands for the model or view and model directly manages the data, logic and rules of the application. In angularjs we consider service is a model. This function is doing the business logic as well, we need to move it into a service function. The function name also doesn't express what it is doing.

gungdesurya1 replied:
ok thanks
should i change `index.js` file to `index.controller.js` too?

gungdesurya1 replied:
@ellyxc I fixed it
please review it again

Note: the `onChangeAcn` function is listening to the ACN input box changes (on key up). And now everytime this function trigger, it will call `isAcnValid` function from service

Commit:
Add Service [ch13440]

- moving the ACN checker algorithm into the service section
- move the target of unit testing to service

ellyxc commented on spec/index.spec.js:
```diff
@@ -0,0 +1,77 @@
+    });
+  }));
+
+  describe("$scope.onChangeAcn", function () {
```
usually we will test the service for this business logic because the service could be used on other controller.
and add another test for controller to test different logic for the controller itself. This PR is looking really good. This sandbox project main goal is to learn how to use our dev tools. Probably left this open and let other team to give feedback and apply them on your telflow portal PR 👍
